syntax = "proto3";

import "Model.proto";

message TaskIDRequest {
  uint32 id = 1;
}

service GrpcTaskService {
  rpc addTask (TaskStructure) returns (AddTaskResponse) {}
  rpc addSubtask (AddSubTaskRequest) returns (AddTaskResponse) {}

  rpc deleteTask(TaskIDRequest) returns(RequstTaskResponse){}
  rpc complete(TaskIDRequest) returns(RequstTaskResponse){}
  rpc postponeTask(PostponeRequest) returns(RequstTaskResponse){}
  rpc editTask(EditRequest) returns(RequstTaskResponse){}

  rpc save(PersistentRequest) returns(RequstTaskResponse){}
  rpc load(PersistentRequest) returns(RequstTaskResponse){}
  
  rpc getTaskByID (TaskIDRequest) returns (GetTaskByIDResponse) {}

  rpc getSubtasks(TaskIDRequest) returns(GetTasksResponse){}
  rpc getAllTasksByPriority(EmptyRequest) returns(GetTasksResponse){}
  rpc getTasksForToday(EmptyRequest) returns(GetTasksResponse){}
  rpc getTasksForWeek(EmptyRequest)returns(GetTasksResponse){}
}

message EmptyRequest{}

message AddSubTaskRequest {
  TaskIDRequest parent = 1;
  TaskStructure task = 2;
}

message PostponeRequest {
  TaskIDRequest id = 1;
  int32 date = 2;
}

message EditRequest {
  TaskIDRequest id = 1;
  TaskStructure task = 2;
}

message PersistentRequest {
  string filename = 1;
}

message AddTaskResponse {
  bool success = 1;
  TaskIDRequest created_id = 2;
  string error_message = 3;
}

message RequstTaskResponse{
    bool success = 1;
    string error_message = 2;
}

message GetTaskByIDResponse {
  TaskStructure task = 1;
}

message GetTasksResponse {
  repeated TaskStructure tasks = 2;
}
